src/MongoDB.Driver.Core/Core/Compression/Zstandard/ZstandardStream.cs

```diff
@@ -0,0 +1,376 @@
+        {
+            throw new NotImplementedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotImplementedException();
+        }
+
+        // private methods
+        private int EnsureCompressionLevelIsValid(Optional<int> compressionLevel)
+        {
+            if (compressionLevel.HasValue && _compressionMode != CompressionMode.Compress)
+            {
+                throw new ArgumentException("Compression level can be specified only in Compress mode", nameof(compressionLevel));
+            }
+
+            if (compressionLevel.HasValue && (compressionLevel.Value <= 0 || compressionLevel.Value > ZstandardNativeWrapper.MaxCompressionLevel))
+            {
+                throw new ArgumentOutOfRangeException(nameof(compressionLevel));
+            }
+
+            return compressionLevel.WithDefault(_defaultCompressionLevel);
+        }
+
+        private CompressionMode EnsureCompressionModeIsValid(CompressionMode compressionMode)
+        {
+            if (compressionMode != CompressionMode.Compress && compressionMode != CompressionMode.Decompress)
+            {
+                throw new ArgumentException($"Invalid compression mode {compressionMode}.", nameof(compressionMode));
+            }
+            return compressionMode;
+        }
+
+        // nested types
+        private class StreamReadHelper
+        {
+            private readonly byte[] _compressedDataBuffer;
+            private readonly Stream _compressedStream;
+            private readonly ReadingState _readingState;
+
+            public StreamReadHelper(
+                Stream compressedStream,
+                byte[] compressedBuffer)
+            {
+                _compressedDataBuffer = Ensure.IsNotNull(compressedBuffer, nameof(compressedBuffer));
+                _compressedStream = Ensure.IsNotNull(compressedStream, nameof(compressedStream));
+                _readingState = new ReadingState();
+            }
+
+            public BufferInfo CompressedBufferInfo => new BufferInfo(_compressedDataBuffer, _readingState.DataPosition);
+
+            // public methods
+            public int ReadCompressedStreamToBufferIfAvailable(
+                int recommendedCompressedSize,
+                out int remainingSize)
+            {
+                remainingSize = _readingState.LastReadDataSize - _readingState.DataPosition;
+
+                if (remainingSize <= 0 && !_readingState.IsDataDepleted && !_readingState.SkipDataReading)
+                {
+                    var readSize = _compressedStream.Read(_compressedDataBuffer, 0, recommendedCompressedSize);
+                    UpdateStateAfterReading(readSize, out remainingSize);
+                }
+
+                return _readingState.DataPosition;
+            }
+
+            public bool TryPrepareDataForNextAttempt(int outputCompressedSize, int outputUncompressedSize)
```

Comment:
For consistency sake name the parameters `compressedBytesProcessed` and `uncompressedBytesProcessed`

Reply:
done